vllm/v1/engine/async_llm.py

```diff
@@ -193,137 +231,93 @@ async def generate(
-
-    def _process_request_outputs(self, request_outputs: List[RequestOutput]):
-        """Process outputs by putting them into per-request AsyncStreams."""
-
-        for request_output in request_outputs:
-            request_id = request_output.request_id
-            assert request_id in self.request_streams
+            )
 
-            # Each request in the API server pulls from the per-request stream.
-            stream = self.request_streams.get(request_id)
-            if stream is not None:
-                stream.put(request_output)
-
-                # If finished, remove from the tracker.
-                if request_output.finished:
-                    if self.log_requests:
-                        logger.info("Finished request %s.", request_id)
-                    self._finish_stream(request_id)
-
-    async def _run_output_handler(self):
-        """Background loop: pulls from EngineCore and pushes to AsyncStreams."""
-
-        try:
+            # The output_handler task pushes items into the queue.
+            # This task pulls from the queue and yields to caller.
             while True:
-                # 1) Pull EngineCoreOutput from the EngineCore.
-                outputs = await self.engine_core.get_output_async()
-
-                # 2) Detokenize based on the output.
-                request_outputs, reqs_to_abort = self.detokenizer.step(outputs)
+                # Note: drain queue without await if possible (avoids
+                # task switching under load which helps performance).
+                out = q.get_nowait() if q.qsize() > 0 else await q.get()
+
+                # Note: both Detokenizer and EngineCore handle their
+                # own request cleanup based on finished.
+                if out.finished:
+                    del self.rid_to_queue[request_id]
+                    yield out
+                    break
+
+                yield out
+
+        # Client request cancellation is handled through calling
+        # task.cancel() on generate(). Calling self.abort() forwards the
+        # cancellation to the EngineCore and Detokenizer.
+        except asyncio.CancelledError:
+            await self.abort(request_id)
+            raise
+
+    async def output_handler_loop(self):
+        """Background loop: pulls from Detokenizer and push to Queues."""
+
+        while True:
+            # Note: use socket directly to avoid calling await multiple
+            # times, which causes too much task switching at high QPS.
+            outputs: List[RequestOutput] = []
+            outputs = await self.from_detokenizer.recv_pyobj()
+
+            for out in outputs:
+                # Note: it is possible that a request was aborted
+                # due to client cancellation while EngineCoreOutputs
+                # are still flowing, so we just ignore.
+                if out.request_id in self.rid_to_queue:
+                    self.rid_to_queue[out.request_id].put_nowait(out)
```

Comment:
Do you think this isn't worth logging since we are sure the request was aborted?